### DISCUSSION TITLE
Keycloak.x custom (federation) provider use datasource in keycloak.conf  quarkus.properties

### DISCUSSION THREAD

**Discussion by simin-c** [keycloak/keycloak#11464] [Archive]
Hi all,

I have a provider that should be migrated from JBOSS distribution to Quarkus distribution.

In the JBOSS distribution I was leveraging the JNDI datasource, while this is becoming vague in the quarkus distribution.

In quarkus.properties I have 

` 
    quarkus.datasource.db-kind=oracle
    quarkus.datasource.jdbc.url=VALIDURL
    quarkus.datasource.username=VALIDUSERNAME
    quarkus.datasource.password=VALIDPASSWORD
    quarkus.datasource.jdbc.max-size=13

`

In build.gradle I have 

`

    // https://mvnrepository.com/artifact/javax.inject/javax.inject
    implementation group: 'javax.inject', name: 'javax.inject', version: '1'

    // https://mvnrepository.com/artifact/io.agroal/agroal-api
    implementation group: 'io.agroal', name: 'agroal-api', version: '1.16'

    // https://mvnrepository.com/artifact/io.agroal/agroal-pool
    implementation group: 'io.agroal', name: 'agroal-pool', version: '1.16'

    // https://mvnrepository.com/artifact/com.oracle.database.jdbc/ojdbc10
    implementation group: 'com.oracle.database.jdbc', name: 'ojdbc10', version: '19.14.0.0'

`

In provider factory I have 

`

     @Inject
     AgroalDataSource dataSource;

`

However on datasource.getConnection() I get 

`

    2022-04-19 10:53:30,473 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-3) Uncaught server error: 
    java.lang.NullPointerException

`

Does anyone has idea here how should the datasources used correctly?

**Comment by steff333** [comment]
Any new progress? I'm facing the same issue and can't access the datasources as shown in 
https://github.com/keycloak/keycloak/blob/main/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/services/health/KeycloakReadyHealthCheck.java#L50

**Reply by simin-c** [reply]
Hi steff333,

Yes, i was able to solve this issue. The key is to use the JPA API, in a glance

import javax.persistence.EntityManager;

       protected EntityManager em = session.getProvider(JpaConnectionProvider.class).getEntityManager();
        this.query = this.em.createNamedQuery("NamedQuery", YOURENTITY.class);

Would do it. 
But you will need to map the relevant table to an entity, and use named query, in addition, scoped import.

Hope this helps!

**Reply by steff333** [reply]
Thanks for the answer.
Anouther way would be to use

javax.enterprise.inject.spi.CDI.current().select(DataSource.class).get()

My Problem was getting a second datasource, no idea why the annotation don't work.
For anyone else running into this issue, there is a chance to get these with

Arc.container().instance(DataSources.class, new Annotation[0]).get()).getDataSource("insert-DS-Name-here")

**Reply by eggyknap** [reply]
Thanks much, that worked great for me. I created `quarkus.properties` in Keycloak's `conf/` directory, defined a new datasource inside (`quarkus.datasource.myname.db-kind=postgresql`, etc.), rebuilt Keycloak, and then from within my user storage provider, could retrieve a datasource like this:
```java
import io.agroal.api.AgroalDataSource;
import io.quarkus.agroal.DataSource;
import io.quarkus.agroal.runtime.DataSources;
import io.quarkus.arc.Arc;
import java.lang.annotation.Annotation;
...
    public void someFunction(...) {
            agroalDataSource = Arc.container().instance(DataSources.class, new Annotation[0]).get().getDataSource("myname");                                                                                      
            conn = agroalDataSource.getConnection();
            // Now use the connection like any other JDBC connection
            // ...
    }
```
The hardest part was probably finding just the right stuff to import :)

**Reply by steff333** [reply]
Hey eggyknap

Great, that its working for you as well. 
I found this solution rather ugly, so after checking the quarkus sourcecode it can be simplified by

```
import io.quarkus.agroal.runtime.DataSources;
....

    private javax.sql.DataSource createDatasource() {
        try {
            return (javax.sql.DataSource) DataSources.fromName("myName");
        }
```

**Reply by eggyknap** [reply]
That's considerably cleaner, thanks. Now, however, I get errors that my connection is already closed, when the enclosing JTA system tries to commit the transaction. This is the latest in a long series of disappointments trying to get this to work; for now I'm going back to our old, soon-to-be-unsupported authentication method, because I've got to get other stuff done than just messing around with Keycloak

**Reply by simin-c** [reply]
> 

Hi, as a reference: I did not use Agoral data source, as that requires an extra import. 

I used this one 

session.getProvider(JpaConnectionProvider.class).getEntityManager();

where the session is the KeycloakSession object in the constructor.

and then 

this.query = this.em.createNamedQuery("QUERY_NAME", YOUREntity.class);

Hope this helps

**Reply by steff333** [reply]
Hey Joshua
even if its sounds a little weird, but I was facing the same issue with transactions.
My first try was defining the keycloak datasource (postgres) via environment variables as in every tutorial and add my second datasource (oracle db) with the conf file. This works ok but I ran in the transaction issues even though I don't use transaction. The oracle db is a second userfederation so read-only access.
Once I defined the postgres DB AND the oracle DB in the conf directory AND define the same postgres connection in the environment variables everythings worked as a charm.
Its a little weird, keycloak uses the connection to postgres (in my case) out of the conf directory during runtime and the connection in the environment variables to create the keycloak DB on the first startup.

Maybe something to try once you have a little time to experiment.

The solution from Simin works as well but only to get a connection the same database that keycloak is using for its own data imho
